cgsmiles/sample.py

```diff
@@ -106,30 +117,58 @@ def grow_chain(self, molecule, seed=None):
         """
         # 1. get the probabilties of any bonding descriptor on the chain to
         #    form the new bond
-        probs = np.array([self.bonding_probabilities[bond_type[:-1]] for bond_type in self.current_open_bonds])
+        probs = np.array([bonding_probabilities[bond_type[:-1]] for bond_type in open_bonds])
         probs = probs / sum(probs)
         # 2. pick a random bonding descriptor according to these probs
-        bonding = np.random.choice(list(self.current_open_bonds.keys()), p=probs)
+        bonding = random.choices(list(open_bonds.keys()), weights=probs)[0]
         # 3. get a corresponding node; it may be that one descriptor is found on
         #    several nodes
-        random.seed(a=seed)
-        source_node = random.choice(self.current_open_bonds[bonding])
+        source_node = random.choice(open_bonds[bonding])
         # 4. get the complementary matching bonding descriptor
         compl_bonding = find_complementary_bonding_descriptor(bonding)
         # 5. pick a new fragment that has such bonding descriptor
-        random.seed(a=seed)
-        fragname, target_node = random.choice(self.fragments_by_bonding[compl_bonding])
+        fragname, target_node = random.choice(fragments[compl_bonding])
         # 6. add the new fragment and do some book-keeping
         correspondence = merge_graphs(molecule, self.fragment_dict[fragname])
         molecule.add_edge(source_node,
                           correspondence[target_node],
-                          bonding=(bonding, compl_bonding))
+                          bonding=(bonding, compl_bonding),
+                          order = int(bonding[-1]))
         molecule.nodes[source_node]['bonding'].remove(bonding)
         molecule.nodes[correspondence[target_node]]['bonding'].remove(compl_bonding)
-        self.current_open_bonds = find_open_bonds(molecule)
         return molecule, fragname
 
-    def terminate_branch(self, molecule, fragname, fragid, seed=None):
+    def terminate_fragment(self, molecule, fragid):
+        """
+        If bonding probabilities for terminal residues are given
+        select one terminal to add to the given fragment. If no
+        terminal bonding probabilities are defined the active bonding
+        descriptors of all nodes will be removed.
+
+        Parameters
+        ----------
+        molecule: nx.Graph
+            the molecule graph
+        fragid: int
+            the id of the fragment
+        """
+        target_nodes = [node for node in molecule.nodes if fragid in molecule.nodes[node]['fragid']]
```

Comment:
```suggestion
        target_nodes = {node for node in molecule.nodes if fragid in molecule.nodes[node]['fragid']}
```
A set should be marginally faster